STORM-1700 Introduce 'whitelist' / 'blacklist' option to MetricsConsumer

HeartSaVioR opened the pull request:
> NOTE: This is based on STORM-1698, PR #1322 

* Users can set whitelist or blacklist to filter out metrics by name
  * if none of them specified (by default), no metrics are filtered out
* how to match: substring match with regular expression
  * use both ^ and $ when you want to match strictly (full string match)
* added unit test

I originally worked based on 1.x-branch and cherry-picked to other branch based on master. 
During cherry-picking, though I remove the changeset of common.clj, git said this file is modified.
If we want to remove this change, I'll add commit to fix this.

HeartSaVioR commented:
Please note that it also opens the possibility to make metrics filters plug-in.
Since downside of making it plug-in is that configuration will be a bit complicated, and I don't see other filters for now so I don't make it plug-in.
But fixing it should be fairly easy, so if anyone can see the benefit please let me know.

HeartSaVioR commented:
I'll rebase before merging.

unsleepy22 commented on storm-core/src/jvm/org/apache/storm/metric/filter/FilterByMetricName.java:
```diff
@@ -0,0 +1,93 @@
+
+        String metricName = dataPoint.name;
+
+        if (!whitelistPattern.isEmpty()) {
```
My concern about pattern match is, regular expression matching is pretty CPU-intensive, if possible, I would prefer a simple "String.contains" match, which may be much faster and less CPU-intensive.

HeartSaVioR replied:
@unsleepy22 
Sorry I just saw this late. Since String.contains could incur unintended matching, I'm in favor of prefix / postfix / infix. Or I suggest alternative way, "keep it as is (using regex) but store matching result into cache" since metrics name is not changed during execution of topology so kinds of names are small.

unsleepy22 replied:
OK, I personally prefer the latter approach.

unsleepy22 commented:
Also, as addressed in STORM-1698, we may have to drop metric points when metrics consumer itself fails to catch up.

HeartSaVioR commented:
For now I applied caching the filter result to avoid determine filter every time.

HeartSaVioR commented:
@unsleepy22 Addressed "dropping older metrics points when consumer can't catch up" to STORM-1698, and rebase this pull request.

unsleepy22 commented on storm-core/src/jvm/org/apache/storm/metric/MetricsConsumerBolt.java:
```diff
@@ -47,17 +68,71 @@ public void prepare(Map stormConf, TopologyContext context, OutputCollector coll
-        _metricsConsumer.handleDataPoints((IMetricsConsumer.TaskInfo)input.getValue(0), (Collection)input.getValue(1));
+        // remove older tasks if task queue exceeds the max size
+        if (_taskQueue.size() > _maxRetainMetricTuples) {
+            while (_taskQueue.size() - 1 > _maxRetainMetricTuples) {
```
I guess we may run into a concurrent issue here, if while statement suffices, it poll one item, while at the same time, metrics consumer consumers one item, anyway, I don't think this matters too much so I'm ok to keep it as is.

HeartSaVioR replied:
I guess maybe it could drop one more item but it's not a big deal anyway. Ideally, taskQueue exceeding the max retain metric tuples should not occur.

abhishekagarwal87 replied:
can we initialize the taskQueue with the capacity? This code will change to
```
while (! _taskQueue.offer(metricTask))
     _taskQueue.poll()
```

HeartSaVioR replied:
Sure, missed while applying restriction of queue size. I'll address it.

abhishekagarwal87 commented on storm-core/src/clj/org/apache/storm/daemon/common.clj:
```diff
@@ -42,17 +42,17 @@
 
 (defmacro defserverfn [name & body]
```
does this class contain only formatting changes?

HeartSaVioR replied:
@abhishekagarwal87 
I didn't intend but it's better to fix it. I'll address it.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/metric/MetricsConsumerBolt.java:
```diff
@@ -47,17 +68,71 @@ public void prepare(Map stormConf, TopologyContext context, OutputCollector coll
     }
-    
+
+    class MetricsTask {
```
This could be a static class.

HeartSaVioR replied:
Yes I think that would be clearer. I'll address it.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/metric/MetricsConsumerBolt.java:
```diff
@@ -47,17 +68,71 @@ public void prepare(Map stormConf, TopologyContext context, OutputCollector coll
+        }
+
+        IMetricsConsumer.TaskInfo taskInfo = (IMetricsConsumer.TaskInfo) input.getValue(0);
+        Collection<IMetricsConsumer.DataPoint> dataPoints = (Collection)input.getValue(1);
```
Minor. space after (Collection)

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/metric/MetricsConsumerBolt.java:
```diff
@@ -17,24 +17,45 @@
-    public MetricsConsumerBolt(String consumerClassName, Object registrationArgument) {
         _consumerClassName = consumerClassName;
         _registrationArgument = registrationArgument;
+        _maxRetainMetricTuples = maxRetainMetricTuples;
```
we can put some sanity check on this argument. ( > 0)

HeartSaVioR replied:
OK. We could even just set queue unbounded when it's lower than 0. I'll address it.

abhishekagarwal87 replied:
Sounds good given that default value is a positive number.

HeartSaVioR replied:
Sure, and behavior should be documented. I'll address it.

abhishekagarwal87 commented on conf/storm.yaml.example:
```diff
@@ -41,8 +41,10 @@
+#     max.retain.metric.tuples: 100
 #     parallelism.hint: 1
 #   - class: "org.mycompany.MyMetricsConsumer"
+#     max.retain.metric.tuples: 100
```
duplicate entry

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/StormCommon.java:
```diff
@@ -383,10 +384,16 @@ public static void addEventLogger(Map conf, StormTopology topology) {
             for (Map<String, Object> info : registerInfo) {
                 String className = (String) info.get("class");
                 Object argument = info.get("argument");
+                Integer maxRetainMetricTuples = Utils.getInt(info.get("max.retain.metric.tuples"), 100);
```
let the default come from defaults.yaml

abhishekagarwal87 replied:
Ahh. it is a per bolt conf. Please ignore the above comment.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/metric/filter/FilterByMetricName.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    private boolean isFilteredIn(String metricName) {
+        if (noneSpecified) {
```
should this be moved/copied to apply function?

HeartSaVioR replied:
For optimization it should be moved at the first line of apply, but logically this check would be good to be close to filtering logic. What do you think?
I don't have strong opinion about this, so I'm also OK to move to apply().

abhishekagarwal87 replied:
I think lots of redundant get/put will be avoided if it is moved.

HeartSaVioR replied:
It avoids lookup with filterCache, but not more than since filter results are cached. Anyway looking up filterCache is also redundant so I'll move it to first line of apply().

HeartSaVioR commented:
NOTE: I'll assume STORM-1698 and STORM-1700 can be merged in this pull request since most of review comments are left here, and addressing comments to STORM-1698 requires STORM-1700 to rebase and resolve conflicts.

HeartSaVioR commented:
Addressed comments.
Actually most of comments are for STORM-1698 but I just set title to STORM-1700 to new commit. It will be mixed up anyway when rebasing so I'll leave as it is.

unsleepy22 commented:
+1

abhishekagarwal87 commented:
Looks good.

HeartSaVioR commented:
I'd like to show result of some performance tests to claim why this improvement is not optional but must.

I modified ThroughputVsLatency to attach addition metrics consumer which sleeps between 0 and 20000 ms (by random) whenever handleDataPoints() is called.

> patch / detach slow consumer

```
uptime:   30 acked:     6,700 acked/sec:     223.33 failed:        0 99%:  19,478,347,775 99.9%:  19,730,006,015 min:   1,031,798,784 max:  20,031,995,903 mean: 9,997,166,638.68 stddev: 4,955,322,338.86 user:    132,350 sys:      6,530 gc:      2,768 mem:     156.95
uptime:   60 acked: 1,250,140 acked/sec:  41,671.33 failed:        0 99%:  30,064,771,071 99.9%:  33,000,783,871 min:       8,007,680 max:  36,339,449,855 mean: 12,109,785,636.33 stddev: 9,009,882,198.75 user:    229,640 sys:     14,950 gc:     48,235 mem:     161.00
uptime:   90 acked:   905,120 acked/sec:  30,170.67 failed:        0 99%:     417,595,391 99.9%:   5,322,571,775 min:       7,536,640 max:   9,546,235,903 mean:   43,809,038.25 stddev:  340,479,608.43 user:    146,070 sys:     21,900 gc:      1,144 mem:     161.92
uptime:  120 acked:   926,660 acked/sec:  30,888.67 failed:        0 99%:      29,409,279 99.9%:      48,463,871 min:       7,438,336 max:      82,771,967 mean:   13,172,560.18 stddev:    3,608,380.15 user:    131,760 sys:     23,500 gc:      1,008 mem:     162.60
uptime:  150 acked:   926,260 acked/sec:  30,875.33 failed:        0 99%:      29,065,215 99.9%:      48,857,087 min:       8,003,584 max:      76,152,831 mean:   13,145,472.87 stddev:    3,538,817.50 user:    128,120 sys:     24,890 gc:        915 mem:     162.90
uptime:  180 acked:   903,140 acked/sec:  30,104.67 failed:        0 99%:      25,624,575 99.9%:      40,566,783 min:       7,962,624 max:      58,785,791 mean:   12,926,076.82 stddev:    2,792,878.03 user:    127,060 sys:     25,820 gc:        869 mem:     163.08
uptime:  210 acked:   903,020 acked/sec:  30,100.67 failed:        0 99%:      30,752,767 99.9%:      45,350,911 min:       7,970,816 max:      73,269,247 mean:   13,114,515.21 stddev:    3,497,139.32 user:    126,060 sys:     26,650 gc:      1,034 mem:     163.30
uptime:  240 acked:   903,160 acked/sec:  30,105.33 failed:        0 99%:      27,541,503 99.9%:      39,976,959 min:       7,921,664 max:      68,091,903 mean:   12,996,252.52 stddev:    3,025,754.06 user:    125,040 sys:     27,630 gc:        968 mem:     163.44
uptime:  270 acked:   902,840 acked/sec:  30,094.67 failed:        0 99%:      28,721,151 99.9%:      45,055,999 min:       7,888,896 max:      71,041,023 mean:   13,062,900.28 stddev:    3,299,611.61 user:    124,240 sys:     28,080 gc:        970 mem:     163.55
uptime:  300 acked:   902,540 acked/sec:  30,084.67 failed:        0 99%:      26,198,015 99.9%:      39,354,367 min:       7,929,856 max:      56,262,655 mean:   12,902,259.50 stddev:    2,817,070.04 user:    124,650 sys:     27,890 gc:        970 mem:     163.61
```

> patch / attach slow consumer

```
uptime:   30 acked:    32,340 acked/sec:   1,078.00 failed:        0 99%:   9,705,619,455 99.9%:   9,923,723,263 min:     724,041,728 max:   9,990,832,127 mean: 8,066,563,909.78 stddev: 1,020,515,611.77 user:    109,100 sys:      4,920 gc:          0 mem:     157.75
uptime:   61 acked: 1,255,480 acked/sec:  40,499.35 failed:        0 99%:  15,879,634,943 99.9%:  17,330,864,127 min:       8,003,584 max:  18,488,492,031 mean: 6,198,566,388.60 stddev: 5,041,039,825.91 user:    191,130 sys:     11,490 gc:     26,684 mem:     162.38
uptime:   91 acked:   904,680 acked/sec:  30,156.00 failed:        0 99%:      33,980,415 99.9%:      60,293,119 min:       7,274,496 max:      84,279,295 mean:   13,399,013.36 stddev:    4,398,380.66 user:    126,980 sys:     17,260 gc:      1,075 mem:     162.93
uptime:  121 acked:   904,780 acked/sec:  30,159.33 failed:        0 99%:      33,718,271 99.9%:      66,519,039 min:       7,598,080 max:     105,644,031 mean:   13,286,131.53 stddev:    4,491,906.80 user:    124,850 sys:     20,730 gc:        971 mem:     163.51
uptime:  151 acked:   903,340 acked/sec:  30,111.33 failed:        0 99%:      26,427,391 99.9%:      48,070,655 min:       7,151,616 max:      82,575,359 mean:   12,966,891.48 stddev:    3,125,613.14 user:    122,320 sys:     22,340 gc:        987 mem:     163.98
uptime:  181 acked:   903,480 acked/sec:  30,116.00 failed:        0 99%:      24,723,455 99.9%:      39,256,063 min:       6,942,720 max:      59,637,759 mean:   12,863,849.51 stddev:    2,673,948.37 user:    121,470 sys:     22,630 gc:        992 mem:     164.32
uptime:  211 acked:   903,040 acked/sec:  30,101.33 failed:        0 99%:      22,773,759 99.9%:      36,896,767 min:       7,835,648 max:      67,764,223 mean:   12,768,220.30 stddev:    2,426,130.42 user:    119,910 sys:     24,270 gc:        933 mem:     164.56
uptime:  241 acked:   903,460 acked/sec:  30,115.33 failed:        0 99%:      25,133,055 99.9%:      57,114,623 min:       7,331,840 max:      92,930,047 mean:   12,921,105.46 stddev:    3,325,266.12 user:    118,350 sys:     24,990 gc:      1,041 mem:     164.68
uptime:  271 acked:   902,980 acked/sec:  30,099.33 failed:        0 99%:      23,822,335 99.9%:      35,454,975 min:       7,913,472 max:      56,066,047 mean:   12,774,768.69 stddev:    2,407,323.94 user:    117,500 sys:     25,880 gc:        940 mem:     164.73
uptime:  301 acked:   903,160 acked/sec:  30,105.33 failed:        0 99%:      24,199,167 99.9%:      40,042,495 min:       6,901,760 max:      76,152,831 mean:   12,811,708.30 stddev:    2,607,966.30 user:    117,830 sys:     25,650 gc:        980 mem:     164.87
```

> before patch / attach slow consumer

```
uptime:   30 acked:    61,380 acked/sec:   2,046.00 failed:        0 99%:   8,824,815,615 99.9%:   9,227,468,799 min:   1,380,974,592 max:   9,487,515,647 mean: 7,127,816,475.13 stddev:  971,681,678.61 user:    110,140 sys:      4,750 gc:          0 mem:     158.50
uptime:   60 acked:   230,940 acked/sec:   7,698.00 failed:        0 99%:  15,854,469,119 99.9%:  16,324,231,167 min:   2,675,965,952 max:  16,634,609,663 mean: 11,353,373,038.74 stddev: 2,644,428,448.90 user:     82,430 sys:      4,050 gc:      6,585 mem:     161.26
uptime:   90 acked:         0 acked/sec:       0.00 failed:        0 99%:               0 99.9%:               0 min:               0 max:               0 mean:            0.00 stddev:            0.00 user:     27,980 sys:      1,700 gc:         38 mem:     161.87
uptime:  120 acked:         0 acked/sec:       0.00 failed:        0 99%:               0 99.9%:               0 min:               0 max:               0 mean:            0.00 stddev:            0.00 user:     30,600 sys:      1,630 gc:         30 mem:     162.48
uptime:  150 acked:         0 acked/sec:       0.00 failed:        0 99%:               0 99.9%:               0 min:               0 max:               0 mean:            0.00 stddev:            0.00 user:     29,600 sys:      1,740 gc:         31 mem:     163.01
uptime:  180 acked:         0 acked/sec:       0.00 failed:        0 99%:               0 99.9%:               0 min:               0 max:               0 mean:            0.00 stddev:            0.00 user:     26,900 sys:      1,990 gc:         21 mem:     163.33
uptime:  210 acked:         0 acked/sec:       0.00 failed:        0 99%:               0 99.9%:               0 min:               0 max:               0 mean:            0.00 stddev:            0.00 user:     27,020 sys:      2,000 gc:         22 mem:     163.69
uptime:  240 acked:         0 acked/sec:       0.00 failed:        0 99%:               0 99.9%:               0 min:               0 max:               0 mean:            0.00 stddev:            0.00 user:     28,060 sys:      2,180 gc:         42 mem:     164.00
uptime:  270 acked:         0 acked/sec:       0.00 failed:        0 99%:               0 99.9%:               0 min:               0 max:               0 mean:            0.00 stddev:            0.00 user:     28,940 sys:      2,370 gc:         22 mem:     164.25
uptime:  300 acked:         0 acked/sec:       0.00 failed:        0 99%:               0 99.9%:               0 min:               0 max:               0 mean:            0.00 stddev:            0.00 user:     28,840 sys:      2,480 gc:         23 mem:     164.40
```

As you may notice, backpressure is triggered from receive queue on metrics consumer, and **it's never back to normal** because metric messages are sent to metrics consumer even though backpressure is on.

![storm-1700-slow-consumer-triggers-backpressure-which-does-not-end](https://cloud.githubusercontent.com/assets/1317309/15239549/61646588-191e-11e6-9aeb-26ebdec9f35c.png)

ptgoetz commented:
See my comments on #1325 as they apply here as well.

harshach commented:
@HeartSaVioR I am leaning towards having a regex metrics filter which users can be speficied in storm.yaml. Instead of pluggable filter implementation. I highly doubt anyone going to write code filter out the metrics although such flexibility will be helpful. How about we go with config option as the first step and if such a need arises we can open up the filter implementation.

HeartSaVioR commented:
@harshach OK. I'm also in doubt regarding this. Let's go with static filter and revisit if necessary.

HeartSaVioR commented:
Addressed Taylor's comment: describe 'how to use' to storm.yaml.example

harshach commented:
+1

ptgoetz commented:
+1

HeartSaVioR commented:
Github fails to track this branch is merged into master. Already merged. Closing.
